File: Examenes/examen1/QytetetRuby/lib/qytetet.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "casilla"
require_relative "sorpresa"
require_relative "dado"
require_relative "jugador"
require_relative "tablero"

require 'singleton'

module ModeloQytetet

  class Qytetet
    
    include Singleton
    
    attr_accessor :jugadorActual, :cartaActual, :jugadores, :tablero, :mazo
    
    @@MAXJUGADORES = 4 # int
    @@MAXCARTAS = 10
    @@MAXCASILLAS = 20
    @@LIBERTAD = 200
    @@SALDOSALIDA = 1000
    
    def initialize
      @jugadores = Array.new #private ArrayList<Jugador>
      nombres = Array.new
      nombres << "juan"
      nombres << "pepe"
      nombres << "luis"
      inicializarJugadores(nombres) #private ArrayList<Jugador>
      @jugadorActual = @jugadores[0] #private Jugador jugadorActual;
      inicializarTablero
      inicializarCartasSorpresa # ArrayList<Sorpresa>
      @cartaActual = @mazo[0] # private Sorpresa cartaActual; // 0 .. 1
      @dado = Dado.instance #private Dado dado = Dado.getInstance();
    end  
    
    def aplicarSorpresa#public boolean aplicarSorpresa()
      raise ArgumentError,'metodo sin implementar'
    end
    
    def cancelarHipoteca(casilla)#public boolean cancelarHipoteca(Casilla casilla)
      raise ArgumentError,'metodo sin implementar'
    end
      
    def comprarTituloPropiedad#public boolean comprarTituloPropiedad()
      raise ArgumentError,'metodo sin implementar'
    end
      
    def edificarCasa(casilla)#public boolean edificarCasa(Casilla casilla)
      raise ArgumentError,'metodo sin implementar'
    end
    
    def edificarHotel(casilla)#public boolean edificarHotel(Casilla casilla)
      raise ArgumentError,'metodo sin implementar'
    end
    
    def hipotecarPropiedad(casilla)#public boolean hipotecarPropiedad(Casilla casilla)
      raise ArgumentError,'metodo sin implementar'
    end
      
    def inicializarJuego(nombres)#public void inicializarJuego( ArrayList<String> nombres)
      raise ArgumentError,'metodo sin implementar'
    end
      
    def intentarSalirCarcel(metodo)#public boolean intentarSalirCarcel( MetodoSalirCarcel metodo)
      raise ArgumentError,'metodo sin implementar'
    end
    
    def jugar #public boolean jugar()
      raise ArgumentError,'metodo sin implementar'
    end
    
    def obtenerRanking #public ArrayList<Jugador> obtenerRanking
      raise ArgumentError,'metodo sin implementar'
    end
    
    def propiedadesHipotecadasJugador(hipotecadas)#public ArrayList<Casilla> propiedadesHipotecadasJugador(boolean hipotecadas)
      raise ArgumentError,'metodo sin implementar'
    end
      
    def siguienteJugador #public Jugador siguienteJugador()
      raise ArgumentError,'metodo sin implementar'
    end
    
    def venderPropiedad(casilla)#public boolean venderPropiedad(Casilla casilla)
      raise ArgumentError,'metodo sin implementar'
    end
    
    private
    
    def encarcelarJugador#private void encarcelarJugador()
      raise ArgumentError,'metodo sin implementar'
    end
    
    def inicializarCartasSorpresa#private void inicializarCartasSorpresa() 
      @mazo = Array.new
      
      @mazo << Sorpresa.new("Tu intuición te dice que mires debajo de esa piedra... ¡Que suerte, 50€!", 50, TipoSorpresa::PAGARCOBRAR)
      @mazo << Sorpresa.new("Tu intuición te dice que mires debajo de esa piedra... Te encuentras a un escarabajo con un cuchillo y te roba 50€",50,TipoSorpresa::PAGARCOBRAR)
      
      @mazo << Sorpresa.new("Te han declarado cómplice de Bárcenas, ¡debes ir a la carcel!", 9, TipoSorpresa::IRACASILLA) ######################## CARCEL #######################
      @mazo << Sorpresa.new("El espacio-tiempo esta hecho un caos, has retrocedido hasta la posición inicial de la partida !y sin cobrar nada¡", 0, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("!Has encontrado un camino subterráneo secreto¡ Avanza hasta la casilla 8.", 8, TipoSorpresa::IRACASILLA)
      
      @mazo << Sorpresa.new("!Todo lo bueno tiene su lado malo¡ tienes que pagar 50€ por cada casa y hotel que tengas para pagar reparaciones.", 50, TipoSorpresa::PORCASAHOTEL)
      @mazo << Sorpresa.new("!Ratas, Ratas por todas partessss¡ Paga 100€ por cada casa y hotel que poseas.", 100, TipoSorpresa::PORCASAHOTEL)
      
      @mazo << Sorpresa.new("¡eres el elegido y por ello cada jugador te da 25€!", 3, TipoSorpresa::PORJUGADOR)
      @mazo << Sorpresa.new("Todo el mundo te mira mal, ¡y con razón! debes pagar 15€ a cada jugador sino quieres que te maten en el siguiente turno ",3,TipoSorpresa::PORJUGADOR)
      
      @mazo << Sorpresa.new("!Es tu dia de suerte, por un error te han declarado inocente y sales de la carcel!",0, TipoSorpresa::SALIRCARCEL)
      
      @mazo.shuffle
      
    end
    
    def inicializarJugadores(nombres)#private void inicializarJugadores(ArrayList<String> nombres ) {if(nombres.size() < MAXJUGADORES && nombres.size() > 2 )
      if (nombres.size() > 2)
        nombres.each do |s|
          if (nombres.size()< @@MAXJUGADORES)
            @jugadores << s
          end
        end
      else
        raise ArgumentError,'numero de jugadores incorrecto'
      end
    end
    
    def inicializarTablero #private void inicializarTablero()
      @tablero = Tablero.new
    end
    
    def salidaJugadores #void salidaJugadores() 
      raise ArgumentError,'metodo sin implementar'
    end
    
    public
    
    def to_s
      "Qytetet:\n
      Jugadores: #{@jugadores}\n
      JugadorActual: #{@jugadorActual}\n
      Tablero: #{@tablero}\n
      Sorpresas: #{@mazo}\n
      CartaActual: #{@cartaActual}\n
      Dado: #{@dado}"
    end
    
  end

end

File: Qytetet_el_juego/p1yp2/QytetetJava/src/modeloqytetet/Qytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.ArrayList;
import static java.util.Collections.shuffle;

/**
 *
 * @author mati
 */

public class Qytetet {
    public static int MAXJUGADORES=4;
    static int MAXCARTAS = 10;
    static int MAXCASILLAS = 20;
    static int LIBERTAD = 200;
    static int SALDOSALIDA = 1000;
    private Dado dado = Dado.getInstance();
    private Sorpresa cartaActual; // 0 .. 1
    private ArrayList<Sorpresa> mazo; //MAX_CARTAS
    private Tablero tablero;
    private Jugador jugadorActual;
    private ArrayList<Jugador> jugadores = new ArrayList();

    
    private static Qytetet instance;
    
    private Qytetet(ArrayList<String> jugadores)throws NumberFormatException{
        inicializarCartasSorpresa();
        inicializarTablero();
        inicializarJugadores(jugadores);
        cartaActual = mazo.get(0);
        jugadorActual = this.jugadores.get(0);
    }

    public ArrayList<Sorpresa> getMazo() {
        return mazo;
    }

    public Tablero getTablero() {
        return tablero;
    }

    public ArrayList<Jugador> getJugadores() {
        return jugadores;
    }
    
    
    
    public static Qytetet getInstance(ArrayList<String> jugadores){
        
        if (instance == null)
            instance = new Qytetet(jugadores);
        return instance;
    }
    /*
    public boolean aplicarSorpresa(){
        return true;
    }
    public boolean cancelarHipoteca(Casilla casilla){
        return true;
    }
    public boolean comprarTituloPropiedad(){
        return true;
    }
    public boolean edificarCasa(Casilla casilla){
        return true;
    }
    public boolean edificarHotel(Casilla casilla){
        return true;
    }
    */
    public Sorpresa getCartaActual(){
        return this.cartaActual;
    }
    public Jugador getJugadorActual(){
        return this.jugadorActual;
    }
    
    /*
    public boolean hipotecarPropiedad(Casilla casilla){
        return true;
    }
    public void inicializarJuego( ArrayList<String> nombres){
    
    } 
    public boolean intentarSalirCarcel( MetodoSalirCarcel metodo) {
        return true;
    } 
    public boolean jugar() {
        return true;
    } 
    public ArrayList<Jugador> obtenerRanking() {
        return null;
    } 
    public ArrayList<Casilla> propiedadesHipotecadasJugador(boolean hipotecadas) {
        return null;
    } 
    public Jugador siguienteJugador() {
        return null;
    }
    
    public boolean venderPropiedad(Casilla casilla) {
        return true;
    } 
    private void encarcelarJugador() {
    
    } 
    */
    private void inicializarCartasSorpresa() {
        mazo = new ArrayList();
      
        mazo.add(new Sorpresa("Tu intuición te dice que mires debajo de esa piedra... ¡Que suerte, 50€!",50,TipoSorpresa.PAGARCOBRAR));
        mazo.add(new Sorpresa("Tu intuición te dice que mires debajo de esa piedra... Te encuentras a un escarabajo con un cuchillo y te roba 50€",50,TipoSorpresa.PAGARCOBRAR));
        
        mazo.add(new Sorpresa("Te han declarado cómplice de Bárcenas, ¡debes ir a la carcel!",9,TipoSorpresa.IRACASILLA)); ///////////////////////
        mazo.add(new Sorpresa("El espacio-tiempo esta hecho un caos, has retrocedido hasta la posición inicial de la partida !y sin cobrar nada¡",0,TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa("!Has encontrado un camino subterráneo secreto¡ Avanza hasta la casilla 8.",8,TipoSorpresa.IRACASILLA));
        
        mazo.add(new Sorpresa("!Todo lo bueno tiene su lado malo¡ tienes que pagar 50€ por cada casa y hotel que tengas para pagar reparaciones.",50,TipoSorpresa.PORCASAHOTEL));
        mazo.add(new Sorpresa("!Ratas, Ratas por todas partessss¡ Paga 100€ por cada casa y hotel que poseas.",100,TipoSorpresa.PORCASAHOTEL));
        
        mazo.add(new Sorpresa("¡eres el elegido y por ello cada jugador te da 25€!",3,TipoSorpresa.PORJUGADOR));
        mazo.add(new Sorpresa("Todo el mundo te mira mal, ¡y con razón! debes pagar 15€ a cada jugador sino quieres que te maten en el siguiente turno ",3,TipoSorpresa.PORJUGADOR));
        
        mazo.add(new Sorpresa("!Es tu dia de suerte, por un error te han declarado inocente y sales de la carcel!",0, TipoSorpresa.SALIRCARCEL));
        
        shuffle(mazo);
    } 
    
    private void inicializarJugadores(ArrayList<String> nombres )throws NumberFormatException{
           this.jugadores = new ArrayList();
           if(nombres.size() < MAXJUGADORES && nombres.size() >= 2 ){
               for(String s:nombres){
                   this.jugadores.add(new Jugador(s));
               }
           }else
              throw new NumberFormatException(" numero de jugadores incorrecto ");
    } 
    
    private void inicializarTablero() {
        this.tablero = new Tablero();// esto llama al constructor y este inicializa.
    } 
    /*
    private void salidaJugadores() {
    
    } 
    */

    @Override
    public String toString() {
        return "Qytetet{" + 
                "dado=" + dado + ",\n "
                + "cartaActual=" + cartaActual + ",\n "
                + "mazo=" + mazo + ", tablero=" + tablero + ",\n "
                + "jugadorActual=" + jugadorActual + ",\n "
                + "jugadores=" + jugadores + '}';
    }
    
    
    
}
